File: frontend/src/lib/triggers.ts
import type { KpiCardData } from "@/components/dashboard/KpiCard";

export type TriggerStatus = "ok" | "watch" | "triggered" | "unknown";

export interface TriggerRule {
  code: string;
  kpiCode: string;
  direction: "above" | "below";
  watchAt: number;
  triggerAt: number;
  /** Skip evaluation when value is outside plausible range (bad source data). */
  validRange?: { min: number; max: number };
}

export const TRIGGER_RULES: TriggerRule[] = [
  {
    code: "trg_unemployment",
    kpiCode: "unemployment",
    direction: "above",
    watchAt: 5,
    triggerAt: 8,
    validRange: { min: 0, max: 100 },
  },
  {
    code: "trg_budget",
    kpiCode: "budget_security",
    direction: "below",
    watchAt: 85,
    triggerAt: 70,
    validRange: { min: 0, max: 200 },
  },
  {
    code: "trg_natural_growth",
    kpiCode: "natural_growth",
    direction: "below",
    watchAt: 0,
    triggerAt: -50,
  },
  {
    code: "trg_doctors",
    kpiCode: "doctors_per_capita",
    direction: "below",
    watchAt: 22,
    triggerAt: 18,
    validRange: { min: 0, max: 100 },
  },
  {
    code: "trg_salary",
    kpiCode: "average_salary",
    direction: "below",
    watchAt: 55_000,
    triggerAt: 45_000,
    validRange: { min: 10_000, max: 500_000 },
  },
  {
    code: "trg_housing",
    kpiCode: "housing_commissioned",
    direction: "below",
    watchAt: 8_000,
    triggerAt: 3_000,
    validRange: { min: 0, max: 1_000_000 },
  },
];

export interface TriggerItem {
  code: string;
  kpi_code: string;
  name: string;
  description: string;
  value: number | null;
  unit: string;
  status: TriggerStatus;
  watch_at: number;
  trigger_at: number;
  direction: "above" | "below";
  is_live: boolean;
}

export interface TriggerLabels {
  name: string;
  description: string;
}

function inRange(value: number, range?: { min: number; max: number }): boolean {
  if (!range) return true;
  return value >= range.min && value <= range.max;
}

export function evaluateTriggerStatus(
  rule: TriggerRule,
  value: number | null | undefined,
): TriggerStatus {
  if (value === null || value === undefined) return "unknown";
  if (!inRange(value, rule.validRange)) return "unknown";

  if (rule.direction === "above") {
    if (value >= rule.triggerAt) return "triggered";
    if (value >= rule.watchAt) return "watch";
    return "ok";
  }

  if (value <= rule.triggerAt) return "triggered";
  if (value <= rule.watchAt) return "watch";
  return "ok";
}

export function buildTriggers(
  kpis: KpiCardData[],
  labels: Record<string, TriggerLabels>,
): TriggerItem[] {
  const byCode = new Map(kpis.map((kpi) => [kpi.code, kpi]));

  return TRIGGER_RULES.map((rule) => {
    const kpi = byCode.get(rule.kpiCode);
    const meta = labels[rule.code];
    return {
      code: rule.code,
      kpi_code: rule.kpiCode,
      name: meta?.name ?? kpi?.name ?? rule.code,
      description: meta?.description ?? "",
      value: kpi?.value ?? null,
      unit: kpi?.unit ?? "—",
      status: evaluateTriggerStatus(rule, kpi?.value),
      watch_at: rule.watchAt,
      trigger_at: rule.triggerAt,
      direction: rule.direction,
      is_live: Boolean(kpi?.is_live && kpi.value !== null),
    };
  });
}

export function countByStatus(triggers: TriggerItem[]): { triggered: number; watch: number; ok: number } {
  return triggers.reduce(
    (acc, item) => {
      if (item.status === "triggered") acc.triggered += 1;
      else if (item.status === "watch") acc.watch += 1;
      else if (item.status === "ok") acc.ok += 1;
      return acc;
    },
    { triggered: 0, watch: 0, ok: 0 },
  );
}

File: backend/app/application/triggers.py
"""Пороговые триггеры по ключевым KPI."""

from dataclasses import dataclass
from typing import Literal

TriggerStatus = Literal["ok", "watch", "triggered", "unknown"]


@dataclass(frozen=True)
class TriggerRule:
    code: str
    kpi_code: str
    direction: Literal["above", "below"]
    watch_at: float
    trigger_at: float
    valid_min: float | None = None
    valid_max: float | None = None


TRIGGER_RULES: tuple[TriggerRule, ...] = (
    TriggerRule("trg_unemployment", "unemployment", "above", 5, 8, 0, 100),
    TriggerRule("trg_budget", "budget_security", "below", 85, 70, 0, 200),
    TriggerRule("trg_natural_growth", "natural_growth", "below", 0, -50),
    TriggerRule("trg_doctors", "doctors_per_capita", "below", 22, 18, 0, 100),
    TriggerRule("trg_salary", "average_salary", "below", 55_000, 45_000, 10_000, 500_000),
    TriggerRule("trg_housing", "housing_commissioned", "below", 8_000, 3_000, 0, 1_000_000),
)

TRIGGER_LABELS: dict[str, tuple[str, str]] = {
    "trg_unemployment": ("Высокая безработица", "Уровень безработицы превышает порог"),
    "trg_budget": ("Бюджетная обеспеченность", "Показатель ниже установленного порога"),
    "trg_natural_growth": ("Естественная убыль", "Отрицательный или низкий прирост населения"),
    "trg_doctors": ("Дефицит врачей", "Обеспеченность медперсоналом ниже нормы"),
    "trg_salary": ("Низкая зарплата", "Средняя зарплата ниже контрольного уровня"),
    "trg_housing": ("Малый ввод жилья", "Объём введённого жилья ниже порога"),
}


def _in_range(value: float, rule: TriggerRule) -> bool:
    if rule.valid_min is not None and value < rule.valid_min:
        return False
    if rule.valid_max is not None and value > rule.valid_max:
        return False
    return True


def evaluate_status(rule: TriggerRule, value: float | None) -> TriggerStatus:
    if value is None:
        return "unknown"
    if not _in_range(value, rule):
        return "unknown"
    if rule.direction == "above":
        if value >= rule.trigger_at:
            return "triggered"
        if value >= rule.watch_at:
            return "watch"
        return "ok"
    if value <= rule.trigger_at:
        return "triggered"
    if value <= rule.watch_at:
        return "watch"
    return "ok"


def build_triggers(kpis: list[dict]) -> list[dict]:
    by_code = {k["code"]: k for k in kpis}
    result: list[dict] = []
    for rule in TRIGGER_RULES:
        kpi = by_code.get(rule.kpi_code, {})
        name, description = TRIGGER_LABELS.get(rule.code, (rule.code, ""))
        value = kpi.get("value")
        is_live = bool(kpi.get("is_live") and value is not None)
        result.append(
            {
                "code": rule.code,
                "kpi_code": rule.kpi_code,
                "name": name,
                "description": description,
                "value": value,
                "unit": kpi.get("unit") or "—",
                "status": evaluate_status(rule, value),
                "watch_at": rule.watch_at,
                "trigger_at": rule.trigger_at,
                "direction": rule.direction,
                "is_live": is_live,
            }
        )
    return result
